http: add perf_hooks detail for http request and client

theanarkh opened the pull request:
http: add perf_hooks detail
   
    1. http: add perf_hooks detail for http request and client
    2. dns,net: move hasObserver out of startPerf and stopPerf

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Affected subsystem: http,net,dns

mscdex commented:
Removing the conditionals is going to unnecessarily make more work on the GC when there are no observers because of the (inline) object being created for every request.

theanarkh commented:
> Removing the conditionals is going to unnecessarily make more work on the GC when there are no observers because of the (inline) object being created for every request.

Thanks for your review. do you mean `hasObserver` ?  I have moved it to `startPerf` and `stopPerf`.

mcollina commented:
Change looks good. We need the benchmarks run.

theanarkh commented:
> Change looks good. We need the benchmarks run.

Thanks for reply,  i wonder what i need to do for `benchmarking` ?

mscdex commented:
> Thanks for your review. do you mean `hasObserver` ? I have moved it to `startPerf` and `stopPerf`.

Yes. Previously object creation was guarded by an `if (hasObserver(..))` but now an object (the one containing `type`, `name`, and `detail`) is created every time whether or not the observer is active.

theanarkh commented:
> 

Thanks for your review. I have modified the code, please review again.

theanarkh commented:
> > Thanks for your review. do you mean `hasObserver` ? I have moved it to `startPerf` and `stopPerf`.
> 
> Yes. Previously object creation was guarded by an `if (hasObserver(..))` but now an object (for example: the one containing `type`, `name`, and `detail`) is created every time whether or not the observer is active.

Thanks for your review. I have modified the code, please review again.

theanarkh commented:
@mcollina @mscdex hi, can you please help review this PR again? thanks.

mcollina commented on lib/internal/perf/observe.js:
```diff
@@ -448,28 +448,27 @@ function hasObserver(type) {
-      startTime: process.hrtime(),
-    };
-  }
+  target[key] = {
```
why this change?

theanarkh replied:
As @mscdex said. 
```
Removing the conditionals is going to unnecessarily make more work on the GC when there are no observers because of the (inline) object being created for every request.
```
So i move the `hasObserver` out of `startPerf` and `stopPerf` to avoid generating useless objects when these are no observer. And the caller need to call  `hasObserver` before call `startPerf` or `stopPerf`

mcollina reviewed (approved):
lgtm

theanarkh commented:
I will separate this PR to two PR.
